# Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# Import data

In [2]:
path = r'C:\Users\dolson\OneDrive - BEACON TECHNOLOGIES\Desktop\Instacart Basket Analysis 10.2022'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
dfords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

# Mixed-Type Data

In [6]:
# Create a dataframe df_test
df_test=pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix']=['a','b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Create custom function to search dataframe for mixed types a for-loop
for col in df_test.columns.tolist():
    weird=(df_test[[col]].applymap(type)!=df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_test[weird]) > 0: 
        print(col)

mix


In [10]:
# Code to convert datatypes .astype('int64')
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [11]:
#Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
#Create new dataframe (subset) to view 16 missing values
df_nan=df_prods[df_prods['product_name'].isnull() == True]

In [13]:
#show rows with missing values
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Misssing Values

In [14]:
# use Mean and df.describe()
#df[column with missings'].fillna(mean value, inplace=True)


In [15]:
# use Median and df_prods.median()
#df['column with missings'].fillna(median value, inplace=True)

In [16]:
#run df_prods.shape function to compare number of rows in current dataframe with number in missing subset after rows removed
df_prods.shape

(49693, 5)

In [17]:
#shape function returns number of rows and columns in a dataframe

In [18]:
# Create new dataframe df_prods_clean
df_prods_clean=df_prods[df_prods['product_name'].isnull()== False]

In [19]:
#check number of rows and columns in new dataframe s/b 16 less rows
df_prods_clean.shape

(49677, 5)

# Duplicates 

In [20]:
# look for full duplicates in dataframe with duplicated function
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
#duplicate rows found in dataframe df_prods_clean
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
# delete duplicates, check current number of rows in df_prods_clean first!
df_prods_clean.shape

(49677, 5)

In [23]:
#create new dataframe that doesn't include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
# Show dataframe count after duplicates deleted
df_prods_clean_no_dups.shape

(49672, 5)

In [25]:
# Save and export new dataframe call 'products_checked.csv' store in prepared data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5 results

In [26]:
# Task2 Run the df.describe() function on your df_prods dataframe.
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [27]:
# Min values appropriate, Max value for prices seems high 

In [28]:
# Task3 check for mixed-type data in df_ords dataframe
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_ords[weird]) > 0: 
        print(col)

In [29]:
# Task 4 no mix datatypes found in df_ords

In [30]:
# Task 5 Run a check for missing values df_ords dataframe
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [31]:
#rename column order_dow to more intuitive name
df_ords.rename(columns={'order_dow' : 'order_day_of_week'} , inplace = True)


In [32]:
#create a dataframe subset to show for the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [33]:
#show rows with missing values
df_ords_nan

order_id  user_id eval_set  order_number  order_day_of_week  \
0         2539329        1    prior             1                  2   
11        2168274        2    prior             1                  2   
26        1374495        3    prior             1                  1   
39        3343014        4    prior             1                  6   
45        2717275        5    prior             1                  3   
...           ...      ...      ...           ...                ...   
3420930    969311   206205    prior             1                  4   
3420934   3189322   206206    prior             1                  3   
3421002   2166133   206207    prior             1                  6   
3421019   2227043   206208    prior             1                  1   
3421069   3154581   206209    prior             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

There are 206,209 missing values in the 'days_since_prior_order' column.
Possible reason for this column's missing values is that these are new customers who have not placed orders
With this high number of missing values it would not be wise to remove them from the analysis
A flag could be created to identify if new_customer (true/false) to tie to missing values

In [34]:
# Task 6 Address missing values using a flag 

In [35]:
# create a new dataframe and set a new column for customer
df_ords_clean = df_ords

In [36]:
# create new flag column 'new_customer' associate to 'days_since_prior_order' column
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [37]:
#look at new dataframe with flag
df_ords_clean

order_id  user_id eval_set  order_number  order_day_of_week  \
0         2539329        1    prior             1                  2   
1         2398795        1    prior             2                  3   
2          473747        1    prior             3                  3   
3         2254736        1    prior             4                  4   
4          431534        1    prior             5                  4   
...           ...      ...      ...           ...                ...   
3421078   2266710   206209    prior            10                  5   
3421079   1854736   206209    prior            11                  4   
3421080    626363   206209    prior            12                  1   
3421081   2977660   206209    prior            13                  1   
3421082    272231   206209    train            14                  6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 8 columns]

In [38]:
# Task 7 Run a check for duplicate values in df_ords dataframe
df_ords_clean[df_ords_clean.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

In [39]:
#Export final cleaned df_prods and df_ords as .csv files to "Prepared Data" folder

There are not duplicates in the df_ords dataframe

In [40]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))

In [41]:
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))